### PR TITLE
enumerable

### PR DESCRIPTION
**Pull request by Marwen-Brini**
in this project i took the different enumerable standard methods such as: each, each_with_index, select, map, any, all, none and inject and i remade them using basic ruby syntax and i even reused some of methods inside other to avoid having to redo the same logic more then once

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
+def my_each(arr)
+  result = []
+  length = arr.length - 1
+  (0..length).each do |i|
+    result.push(yield(arr[i]))
+  end
+  result
+end
```
Comment (pacyL2K19) [inline enumerable.rb:8]: [Mandatory] Please also consider the case the bloc was not given

**Review by pacyL2K19 — CHANGES_REQUESTED** [review]
# Status: Changes Required :stop_sign: 
Hello **Team** you made a great job so far and your code looks clean, I enjoyed reading it
However, there are some changes and improvement you need to make before getting approved 
Please consider these changes as **Mandatory** and work on them before you submit another one 
- [ ] The readme file looks so minimalist feel free to make a professional on following [THIS]() Microverse template 
- [ ] The code many methods don't work as expected and throw an exception on execution see this expectation below on the `my_each` method
warnEnn1
- [ ] Also consider cases when the bloc was not given trying to follow the original Enumerable behavior returning an [Enumerator]() when the bloc is not given 
- [ ] Your project seems to miss some required methods like `my_count` please make sure you created all the required method before another submission
- [ ] When testing your project before submitting it, kindly compare what the original enumerable return with what your custom method return with the same parameters and make sure they always the same like below :arrow_down: 
warnTrueEx

**Comment by rukundoeric** [comment]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve
  - [ ] [MANDATORY] #my_each when a block is given when self is an array returns the array itself after calling the given block once for each element in self.
    - when array is self and block is given. `[1,2,3,4].my_each(proc { |num| num % 2 = 0 })` 
       - Expected: `[1,2,3,4]` 
       - Got: `nil`
    - when range is self and block is given. `Range.new(5, 50).my_each(proc { |num| num % 2 = 0 })` 
       - Expected: `5..50` 
       - Got: `nil`
  - [ ] [MANDATORY] #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self.
     - when array is self and block is given. `[1,2,3,4].my_each(proc { |num| num % 2 = 0 })` 
       - Expected: `[1,2,3,4]` 
       - Got: `nil`
    - when range is self and block is given. `Range.new(5, 50).my_each(proc { |num| num % 2 = 0 })` 
       - Expected: `5..50` 
       - Got: `nil`
  - [ ] [MANDATORY] #my_all? when no block or argument is given returns true when none of the collection members are false or nil.
    - when array is self and no block given ` [1, true, 'hi', []].my_all?`
      - Expected: `true`
      - Got: `false`
  - [ ] [MANDATORY] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer or class::Numeric. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil.
     - when an array is self and no block is given `[nil, false, true, []].my_any?`
       - Expected: `true`
       - Got: `false`
  - [ ] [MANDATORY] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class: Numeric, Integer, etc. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern. In this case, your function is throwing an error.
  - [ ] [MANDATORY] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern In this case your function is throwing an error.
  - [ ] [MANDATORY] #my_none? when no block or argument is given returns true only if none of the collection members is true.
    - when an array is self and no block is given `[nil, false, nil, false].my_none?`
      - Expected: `true`
      - Got: `false`
  - [ ] [MANDATORY] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class: Numeric, Integer, etc. In this case, your function is throwing an error.
     
  - [ ] [OPTIONAL] It would be better to rename your file from enumerable.rb` to `enumerables.rb` so that it will be easier and faster for other TSEs to run and test your functions. 

 NOPE: Almost all methods are failing to pass the test, kindly make sure each of your method works exactly like its original in all cases.  
         


_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      if block_given?
+        return false unless yield each
+      elsif par.instance_of? Class
+        return false unless each.is_a? par
+      else
+        return false unless par.nil? ? each : none_nil?(par, each)
+      end
+    end
+    true
+  end
```
Comment (henatan99) [inline enumerables.rb:39]: *[ ] when a Regex is passed as an argument it returns true if all of the collection matches the Regex, but your my_all? method returns false. Please see the attached screenshot of the test run on a cloned repo of your project.  
Screenshot from 2021-03-27 20-20-23

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      if block_given?
+        return true if yield item
+      elsif par.instance_of? Class
+        return true if item.is_a? par
+      elsif par.nil? ? item : none_nil?(par, item)
+        return true
+      end
+    end
+    false
+  end
```
Comment (henatan99) [inline enumerables.rb:52]: *[ ] when a Regex is passed as an argument it returns true if any of the collection matches the Regex, but your my_any? method returns false. See the attached screenshot to compare the output of your my_any? method with the built-in any? method. 
Screenshot from 2021-03-27 20-25-55

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      if block_given?
+        return false if yield i
+      elsif pam.instance_of? Class
+        return false unless i.is_a? pam
+      elsif pam.nil? ? i == true : none_nil?(pam, i)
+        return false
+      end
+    end
+    true
+  end
```
Comment (henatan99) [inline enumerables.rb:65]: when a Regex is passed as an argument it returns false if any of the collection matches the Regex, but your my_none method returns true. Look at the screenshot attached to compare the output of your my_none method with the built in none method. 
Screenshot from 2021-03-27 20-29-49

**Review by henatan99 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:

Hey Team, you have done a nice job in this project, but you need to fix a few failing cases of some of your enumerable methods 

## To Highlight 
*[ ] Most of the methods work exactly as the built-in ruby enumerable methods 
*[ ] You have done well addressing the requested changes from the previous reviewers 

## To Improve 
*[ ] when a Regex is passed as an argument it returns true if all of the collection matches the Regex, but your my_all? method returns false. Please see the inline comment.

*[ ] when a Regex is passed as an argument it returns true if any of the collection matches the Regex, but your my_any? method returns false. Please see the inline comment.

*[ ]  when a Regex is passed as an argument it returns false if any of the collection matches the Regex, but your my_none method returns true. Please see the inline comment.

*[ ] There is a typo in your PR title 

### Please work on the requested changes and submit again for re-review.

### Happ Coding :computer: :computer_mouse: 

If you want to reach out to me, follow the below links:
- [LinkedIn](https://www.linkedin.com/in/henok-mekonnen-2a251613/)
- [Twitter](https://twitter.com/henatan99)
- Henok Mekonnen on Slack

**Review by Dipeshtwis — APPROVED** [review]
# Status: Approved ✔️
## Hi @Marwen-Brini   and @Abdona. I'm Dipesh your reviewer for this PR 👩‍🚀 

Good job on implementing all the changes requested in the previous review. well done

## Optional yet Highly recommended

- [ ] should have working `my_inject` method identical and returns the same thing as ruby's inject. For now, I can't able to see the `my_inject` method.

- [ ] Fix the readme instruction to run the file. I think it should be `ruby enumerables.rb` instead of `ruby enumerable.rb`.
- [x] There is a typo in your PR title. Please fix it.

From the next Project keep in mind
- [ ] Always make a development branch, change it for default, and do the Pull Request from feature branch to development.


**You may close this Pull Request after making the changes and merge**
---------
    
**Keep being awesome devs and good luck with your next project!**
---------

Cheers and Happy coding!👏👏👏

Code review by **Dipesh Kumar**

(Feel free to contact me on slack by the handle of  **deep** if you have any additional question).
Happy Coding
[Dipesh Kumar](https://deeptwist.tk)
[Github](https://github.com/Dipeshtwis)

**Pull request merged**